File: java/src/kanzi/test/TestMTFT.java
/*
Copyright 2011-2013 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.test;

import java.util.Random;
import kanzi.IndexedByteArray;
import kanzi.transform.MTFT;


public class TestMTFT
{
    public static void main(String[] args)
    {
        System.out.println("TestMTFT");

        // Behavior Test
        {
            System.out.println("\nCorrectness test");

            for (int ii=1; ii<=20; ii++)
            {
                byte[] input;
                
                if (ii == 1)
                {
                    input = new byte[] { 5, 2, 4, 7, 0, 0, 7, 1, 7 };
                }
                else
                {
                    input = new byte[32];
                    Random rnd = new Random();

                    for (int i=0; i<input.length; i++)
                    {
                        input[i] = (byte) (65 + rnd.nextInt(5*ii));
                    }
                }

                int size = input.length;
                MTFT mtft = new MTFT(size);
                byte[] transform = new byte[size+20];
                byte[] reverse = new byte[size];
                IndexedByteArray iba1 = new IndexedByteArray(input, 0);
                IndexedByteArray iba2 = new IndexedByteArray(transform, 0);
                IndexedByteArray iba3 = new IndexedByteArray(reverse, 0);

                System.out.println("\nTest "+ii);
                System.out.print("Input     : ");

                for (int i=0; i<size; i++)
                {
                    System.out.print((input[i] & 0xFF) + " ");
                }

                int start = (ii & 1) * ii;
                iba2.index = start;
                mtft.forward(iba1, iba2);
                System.out.println();
                System.out.print("Transform : ");

                for (int i=start; i<start+size; i++)
                {
                    System.out.print((transform[i] & 0xFF) + " ");
                }

                iba2.index = start;
                mtft.inverse(iba2, iba3);
                System.out.println();
                System.out.print("Reverse   : ");

                for (int i = 0; i < size; i++)
                {
                    System.out.print((reverse[i] & 0xFF) + " ");
                }

                System.out.println();
                boolean ok = true;

                for (int i=0; i<size; i++)
                {
                    if (reverse[i] != input[i])
                    {
                        ok = false;
                        break;
                    }
                }

                System.out.println((ok == true) ? "Identical" : "Different");

                if (ok == false)
                  System.exit(1);
            }
        }

      // Speed Test
      final int iter = 20000;
      final int size = 10000;
      System.out.println("\n\nSpeed test");
      System.out.println("Iterations: "+iter);
     
      for (int jj=0; jj<4; jj++)
      {     
            byte[] input = new byte[size];
            byte[] transform = new byte[size];
            byte[] reverse = new byte[size];
            IndexedByteArray iba1 = new IndexedByteArray(input, 0);
            IndexedByteArray iba2 = new IndexedByteArray(transform, 0);
            IndexedByteArray iba3 = new IndexedByteArray(reverse, 0);
            MTFT mtft = new MTFT();
            long delta1 = 0, delta2 = 0;
            long before, after;

            if (jj == 0)
               System.out.println("\nPurely random input");
            
            if (jj == 2)
               System.out.println("\nSemi random input");
               
            for (int ii = 0; ii < iter; ii++)
            {
                Random rnd = new Random();
                int n = 128;
                
                for (int i = 0; i < input.length; i++)
                {
                    if (jj < 2)
                    {
                       // Pure random
                       input[i] = (byte) (rnd.nextInt(256));
                    }
                    else
                    {
                       // Semi random (a bit more realistic input)
                       int rng = ((i & 7) == 0) ? 256 : 5;
                       int p = (rnd.nextInt(rng) - rng/2 + n) & 0xFF;
                       input[i] = (byte) p;
                       n = p;
                    }
                }

                before = System.nanoTime();
                iba1.index = 0;
                iba2.index = 0;
                mtft.forward(iba1, iba2);
                after = System.nanoTime();
                delta1 += (after - before);
                before = System.nanoTime();
                iba2.index = 0;
                iba3.index = 0;
                mtft.inverse(iba2, iba3);
                after = System.nanoTime();
                delta2 += (after - before);

                int idx = -1;

                // Sanity check
                for (int i=0; i<size; i++)
                {
                   if (input[i] != reverse[i])
                   {
                      idx = i;
                      break;
                   } 
                }

                if (idx >= 0)
                   System.out.println("Failure at index "+idx+" ("+input[idx]+"<->"+reverse[idx]+")");
            }

            final long prod = (long) iter * (long) size;
            System.out.println("MTFT Forward transform [ms]: " + delta1 / 1000000);
            System.out.println("Throughput [KB/s]          : " + prod * 1000000L / delta1 * 1000L / 1024);
            System.out.println("MTFT Reverse transform [ms]: " + delta2 / 1000000);
            System.out.println("Throughput [KB/s]          : " + prod * 1000000L / delta2 * 1000L / 1024);        
            System.out.println();
        }
    }
} 

File: cpp/src/test/TestMTFT.cpp
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <iostream>
#include <fstream>
#include <ctime>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <algorithm>
#include "../types.hpp"
#include "../transform/MTFT.hpp"

using namespace std;
using namespace kanzi;

void testMTFTCorrectness()
{
    // Test behavior
    cout << "MTFT Correctness test" << endl;
    srand(uint(time(nullptr)));

    for (int ii = 0; ii < 20; ii++) {
        byte val[32];
        int size = 32;

        if (ii == 0) {
            byte val2[] = { 5, 2, 4, 7, 0, 0, 7, 1, 7 };
            size = 32;
            memset(&val[0], 0, size);
            memcpy(&val[0], &val2[0], 9);
        }
        else {
            for (int i = 0; i < 32; i++)
                val[i] = byte(65 + (rand() % (5 * ii)));
        }

        MTFT mtft;
        byte* input = &val[0];
        byte* transform = new byte[size + 20];
        byte* reverse = new byte[size];
        cout << endl
             << "Test " << (ii + 1);
        cout << endl
            << "Input     : ";

        for (int i = 0; i < size; i++)
            cout << (input[i] & 0xFF) << " ";

        int start = (ii & 1) * ii;
        SliceArray<byte> ia1(input, size, 0);
        SliceArray<byte> ia2(transform, size + 20, start);
        mtft.forward(ia1, ia2, size);

        cout << endl
             << "Transform : ";

        for (int i = start; i < start + size; i++)
            cout << (transform[i] & 0xFF) << " ";

        SliceArray<byte> ia3(reverse, size, 0);
        ia2._index = start;
        mtft.inverse(ia2, ia3, size);

        bool ok = true;
        cout << endl
           << "Reverse   : ";

        for (int i = 0; i < size; i++)
            cout << (reverse[i] & 0xFF) << " ";

        for (int j = 0; j < size; j++) {
            if (input[j] != reverse[j]) {
                ok = false;
                break;
            }
        }

        cout << endl;
        cout << ((ok) ? "Identical" : "Different") << endl;
        delete[] transform;
        delete[] reverse;
    }
}

int testMTFTSpeed()
{
    // Test speed
    int iter = 20000;
    int size = 10000;
    cout << endl
         << endl
         << "MTFT Speed test" << endl;
    cout << "Iterations: " << iter << endl;
    srand(uint(time(nullptr)));

    for (int jj = 0; jj < 4; jj++) {
        byte input[20000];
        byte output[20000];
        byte reverse[20000];
        MTFT mtft;
        double delta1 = 0, delta2 = 0;

        if (jj == 0) {
            cout << endl
                 << endl
                 << "Purely random input" << endl;
        }

        if (jj == 2) {
            cout << endl
                 << endl
                 << "Semi random input" << endl;
        }

        for (int ii = 0; ii < iter; ii++) {
            for (int i = 0; i < size; i++) {
                int n = 128;

                if (jj < 2) {
                    // Pure random
                    input[i] = byte(rand() % 256);
                }
                else {
                    // Semi random (a bit more realistic input)
                    int rng = 5;

                    if ((i & 7) == 0) {
                        rng = 128;
                    }

                    int p = ((rand() % rng) - rng / 2 + n) & 0xFF;
                    input[i] = (byte)p;
                    n = p;
                }
            }

            SliceArray<byte> ia1(input, size, 0);
            SliceArray<byte> ia2(output, size, 0);
            SliceArray<byte> ia3(reverse, size, 0);
            clock_t before1 = clock();
            mtft.forward(ia1, ia2, size);
            clock_t after1 = clock();
            delta1 += (after1 - before1);
            clock_t before2 = clock();
            ia2._index = 0;
            mtft.inverse(ia2, ia3, size);
            clock_t after2 = clock();
            delta2 += (after2 - before2);

            // Sanity check
            int idx = -1;

            for (int i = 0; i < size; i++) {
                if (input[i] != reverse[i]) {
                    idx = i;
                    break;
                }

                if (idx >= 0) {
                    cout << "Failure at index " << i << " (" << (int)input[i] << "<->" << (int)reverse[i] << ")" << endl;
                    break;
                }
            }
        }

        double prod = double(iter) * double(size);
        double b2KB = double(1) / double(1024);
        double d1_sec = double(delta1) / CLOCKS_PER_SEC;
        double d2_sec = double(delta2) / CLOCKS_PER_SEC;
        cout << "MTFT Forward transform [ms]: " << int(d1_sec * 1000) << endl;
        cout << "Throughput [KB/s]          : " << int(prod * b2KB / d1_sec) << endl;
        cout << "MTFT Reverse transform [ms]: " << int(d2_sec * 1000) << endl;
        cout << "Throughput [KB/s]          : " << int(prod * b2KB / d2_sec) << endl;
        cout << endl;
    }

    return 0;
}

#ifdef __GNUG__
int main(int, const char**)
#else
int TestMTFT_main(int, const char**)
#endif
{
    testMTFTCorrectness();
    testMTFTSpeed();
    return 0;
}
